content/FIDO2/FIDO2_WebAuthn_Developer_Guide/WebAuthn_Server_Registration.adoc

```diff
@@ -0,0 +1,72 @@
+
+== WebAuthn Server Registration
+
+=== Initiate Registration
+To initiate the registration ceremony the application will construct a StartRegistrationOptions instance using its builder and pass that info into the RelyingParty startRegistration method. The only require parameter is a UserIdentity describing the user for which to create a credential. The id field contains the user handle. This should be a stable, unique identifier for the user - equivalent to a username. However due to privacy considerations it is recommended to set the user handle to a random byte array rather than a username.
+
+[source,java]
+----
+byte[] userHandle = new byte[64];
+random.nextBytes(userHandle);
+
+PublicKeyCredentialCreationOptions request = rp.startRegistration(StartRegistrationOptions.builder()
+    .user(UserIdentity.builder()
+        .name("alice")
+        .displayName("Alice Hypothetical")
+        .id(new ByteArray(userHandle))
+        .build())
+    .build());
+----
+
+The startRegistration method returns a PublicKeyCredentialCreationOptions which can be serialized to JSON, sent to the client, and passed as the publicKey argument to navigator.credentials.create(). This request should be stored in temporary storage so that it can later be used in the finish registration step.
+
+[source,java]
+----
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+ObjectMapper jsonMapper = new ObjectMapper()
+    .configure(SerializationFeature.FAIL_ON_EMPTY_BEANS, false)
+    .setSerializationInclusion(Include.NON_ABSENT)
+    .registerModule(new Jdk8Module());
+
+String json = jsonMapper.writeValueAsString(request);
+return json;
+----
+
+=== Finish Registration
+After receiving the response from the client construct a PublicKeyCredential from the response and wrap it in a FinishRegistrationOptions along with the PublicKeyCredentialCreationOptions which was used to initiate the request.  
+
+[source,java]
+----
+String responseJson = /* ... */;
+PublicKeyCredential<AuthenticatorAttestationResponse, ClientRegistrationExtensionOutputs> pkc =
+    jsonMapper.readValue(responseJson, new TypeReference<PublicKeyCredential<AuthenticatorAttestationResponse, ClientRegistrationExtensionOutputs>>(){});
+----
+
+Pass the PublicKeyCredential as the argument to the RelyingParty finishRegistration method to validate the response. Be sure to remove the PubicKeyCredentialCreationOptions from the pending requests storage to prevent retries. The library performs a series of checks to ensure the registration ceremony was not tampered with. Including:
+Verifying the challenge is the same as the challenge that was sent
+Ensuring the origin was the origin expected
+Validating that the signature over the clientDataHash and the attestation using the certificate chain for that specific model of the authenticator.
```

Comment:
List formatting was lost in migration to adoc

Reply:
incorporated into commit eacd866